odl/space/fspace.py

```diff
@@ -490,29 +908,161 @@ def _multiply(x, y):
         The multiplication is implemented with a simple Python
         function, so the resulting function object is probably slow.
         """
-        x_old = x.function
-        y_old = y.function
+        x1_call_oop = x1._call_out_of_place
+        x1_call_ip = x1._call_in_place
+        x2_call_oop = x2._call_out_of_place
+        x2_call_ip = x2._call_in_place
+
+        product_vect = x1.vectorized or x2.vectorized
+        dtype = complex if self.field == ComplexNumbers() else float
+        # Manually vectorize if necessary. Use out-of-place for both
+        if product_vect and not x1.vectorized:
+            x1_call_oop = vectorize(dtype, outarg='none')(x1_call_oop)
+            x1_call_ip = vectorize(dtype, outarg='positional')(x1_call_oop)
+        if product_vect and not x2.vectorized:
+            x2_call_oop = vectorize(dtype, outarg='none')(x2_call_oop)
+            x2_call_ip = vectorize(dtype, outarg='positional')(x2_call_oop)
+
+        def product_call_out_of_place(x):
+            """The product out-of-place evaluation function."""
+            return x1_call_oop(x) * x2_call_oop(x)
+
+        def product_call_in_place(x, out):
+            """The product in-place evaluation function."""
+            tmp = np.empty_like(out)
+            x1_call_ip(x, out)
+            x2_call_ip(x, tmp)
+            out *= tmp
+            return out
 
-        def product(arg):
-            """The actual product function."""
-            return x_old(arg) * y_old(arg)
-        y._function = product
+        def product_call(x, out=None):
+            """Product, dual-use version for final use."""
+            if out is None:
+                return product_call_out_of_place(x)
+            else:
+                return product_call_in_place(x, out)
 
-    def __repr__(self):
-        """Return ``repr(self)``."""
-        inner_repr = '{!r}'.format(self.domain)
-        if not isinstance(self.range, RealNumbers):
-            inner_repr += ', {!r}'.format(self.range)
+        if product_vect:
+            out._call_out_of_place = out._call_in_place = product_call
+            out._call_has_out = out._call_out_optional = True
+        else:
+            out._call_out_of_place = product_call_out_of_place
+            out._call_in_place = _default_in_place
+            out._call_has_out = out._call_out_optional = False
 
-        return 'FunctionSpace({})'.format(inner_repr)
+        out._vectorized = product_vect
 
-    def __str__(self):
-        """Return ``str(self)``."""
-        inner_repr = '{}'.format(self.domain)
-        if not isinstance(self.range, RealNumbers):
-            inner_repr += ', {}'.format(self.range)
+        return out
+
+    def _divide(self, x1, x2, out):
```

Comment:
Exactly the same as multiply except about 3 characters, has to be able to factor out the common code?